core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultPendingLockManager.java

```diff
@@ -0,0 +1,381 @@
+      }
+
+      return timeService.remainingTime(expectedEndTime, TimeUnit.MILLISECONDS);
+   }
+
+   private long checkForAnyPendingLocks(Collection<Object> keys, GlobalTransaction globalTransaction, int transactionTopologyId, long lockTimeout) throws InterruptedException {
+      if (trace)
+         log.tracef("Checking for pending locks and then locking key %s", toStr(keys));
+
+      final long expectedEndTime = timeService.expectedEndTime(lockTimeout, TimeUnit.MILLISECONDS);
+
+      final Collection<PendingTransaction> pendingTransactions = getTransactionWithAnyLockedKey(transactionTopologyId, keys, globalTransaction);
+      final PendingTransaction lockOwner = waitForTransactionsToComplete(pendingTransactions, expectedEndTime);
+
+      // Then try to acquire a lock
+      if (trace) {
+         log.tracef("Finished waiting for other potential lockers. Timed-Out? %b", lockOwner != null);
+      }
+
+      if (lockOwner != null) {
+         timeout(lockOwner, globalTransaction);
+      }
+
+      return timeService.remainingTime(expectedEndTime, TimeUnit.MILLISECONDS);
+   }
+
+   private void timeout(PendingTransaction lockOwner, GlobalTransaction thisGlobalTransaction) {
+      throw new TimeoutException(format("Could not acquire lock on %s in behalf of transaction %s. Current owner %s.",
+                                        lockOwner.commonKey, thisGlobalTransaction,
+                                        lockOwner.cacheTransaction.getGlobalTransaction()));
+   }
+
+   private PendingTransaction waitForTransactionsToComplete(Collection<PendingTransaction> transactionsToCheck,
+                                                            long expectedEndTime) throws InterruptedException {
+      if (transactionsToCheck.isEmpty()) {
+         return null;
+      }
+      for (PendingTransaction tx : transactionsToCheck) {
+         long remaining;
+         if ((remaining = timeService.remainingTime(expectedEndTime, TimeUnit.MILLISECONDS)) > 0) {
+            if (!tx.cacheTransaction.waitForLockRelease(remaining)) {
+               return tx;
+            }
+         }
+      }
+      return null;
+   }
+
+   private Collection<PendingTransaction> getTransactionWithLockedKey(int transactionTopologyId,
+                                                                      Object key,
+                                                                      GlobalTransaction globalTransaction) {
+      final Collection<PendingTransaction> pendingTransactions = new ArrayList<>();
+      forEachTransaction(transaction -> {
+         if (transaction.getTopologyId() < transactionTopologyId &&
+               !transaction.getGlobalTransaction().equals(globalTransaction) &&
+               transaction.containsLockOrBackupLock(key)) {
+            pendingTransactions.add(new PendingTransaction(transaction, key));
+         }
+      });
+      return pendingTransactions.isEmpty() ? Collections.emptyList() : pendingTransactions;
```

Comment:
I don't think the conditional is needed, the `pendingTransactions` list is already instantiated, so using `emptyList()` doesn't save us the allocation.